cmd/skopeo/sync.go

```diff
@@ -0,0 +1,594 @@
+
+	destinationURL, err := parseURL(args[1])
+	if err != nil {
+		return errors.WithMessage(err, "Error while parsing destination")
+	}
+	destinationCtx, err := opts.destImage.newSystemContext()
+	if err != nil {
+		return err
+	}
+
+	sourceCtx, err := opts.srcImage.newSystemContext()
+	if err != nil {
+		return err
+	}
+	sourceArg := args[0]
+
+	var srcRepoList []repoDescriptor
+
+	if opts.sourceYaml {
+		srcRepoList, err = syncFromYaml(sourceArg, sourceCtx)
+		if err != nil {
+			return err
+		}
+	} else {
+		sourceURL, err := parseURL(sourceArg)
+		if err != nil {
+			return errors.WithMessage(err, "Error while parsing source")
+		}
+
+		if transports.Get(sourceURL.Scheme) == directory.Transport &&
+			sourceURL.Scheme == destinationURL.Scheme {
+			return errors.New("sync from 'dir:' to 'dir:' not implemented, use something like rsync instead")
+		}
+
+		srcRepo, err := syncFromURL(sourceURL, sourceCtx)
+		if err != nil {
+			return err
+		}
+		srcRepoList = append(srcRepoList, srcRepo)
+	}
+
+	ctx, cancel := opts.global.commandTimeoutContext()
+	defer cancel()
+
+	var imgCounter int
+	for _, srcRepo := range srcRepoList {
+		options := copy.Options{
+			RemoveSignatures: opts.removeSignatures,
+			SignBy:           opts.signByFingerprint,
+			ReportWriter:     os.Stdout,
+			DestinationCtx:   destinationCtx,
+			SourceCtx:        srcRepo.Context,
+		}
+
+		for counter, ref := range srcRepo.TaggedImages {
+			destRef, err := buildFinalDestination(ref, destinationURL, srcRepo.DirBasePath)
+			if err != nil {
+				return err
+			}
+
+			logrus.WithFields(logrus.Fields{
+				"from": transports.ImageName(ref),
+				"to":   transports.ImageName(destRef),
+			}).Infof("Copying image tag %d/%d", counter+1, len(srcRepo.TaggedImages))
+
+			_, err = copy.Image(ctx, policyContext, destRef, ref, &options)
+			if err != nil {
+				return errors.WithMessage(err, fmt.Sprintf("Error copying tag '%s'", transports.ImageName(ref)))
+			}
+			imgCounter++
+		}
+	}
+
+	logrus.Infof("Synced %d images from %d sources", imgCounter, len(srcRepoList))
+
+	return nil
+}
+
+func syncCmd(global *globalOptions) cli.Command {
+	sharedFlags, sharedOpts := sharedImageFlags()
+	srcFlags, srcOpts := imageFlags(global, sharedOpts, "src-", "screds")
+	destFlags, destOpts := imageDestFlags(global, sharedOpts, "dest-", "dcreds")
+	opts := syncOptions{global: global,
+		srcImage:  srcOpts,
+		destImage: destOpts,
+	}
+
+	filterFlags := func(flags []cli.Flag, prefix string) []cli.Flag {
+		flagsNotNeeded := []string{
+			"daemon-host",
+			"ostree-tmp-dir",
+			"shared-blob-dir",
+		}
+
+		filtered := flags[:0]
+		for _, f := range flags {
+			var found bool
+			for _, e := range flagsNotNeeded {
+				if e == strings.TrimPrefix(f.GetName(), prefix) {
+					found = true
+					break
+				}
+			}
+			if !found {
+				filtered = append(filtered, f)
+			}
+		}
+		return filtered
+	}
+
+	srcFlags = filterFlags(srcFlags, "src-")
+	destFlags = filterFlags(destFlags, "dest-")
```

Comment:
This feels unclean, I’ll take another look.

Reply:
This is required because some of the `imageOptions` and `destImageOptions` are not meaningful to the sync command. 
So I am filtering those options out. I agree this is sub-optimal, we could find a better way.

Reply:
We could split `imageOptions` by transports, or pass something like `skipTransports []string` to `imageFlags`—but it’s very tempting to just not leave the options in and not do any filtering.

Let’s first figure out how to specify/parse the hierarchy roots; that might change how we refer to transports.